pkg/hubble/relay/peer.go

```diff
@@ -0,0 +1,146 @@
+
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+)
+
+func newPeerClient(target string, dialTimeout time.Duration) (peerpb.PeerClient, *grpc.ClientConn, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), dialTimeout)
+	defer cancel()
+	// the connection is assumed to be local
+	conn, err := grpc.DialContext(ctx, target, grpc.WithInsecure(), grpc.WithBlock())
+	if err != nil {
+		return nil, nil, err
+	}
+	return peerpb.NewPeerClient(conn), conn, nil
+}
+
+func (s *Server) syncPeers() {
+	ctx, cancel := context.WithCancel(context.Background())
+connect:
+	for {
+		cl, conn, err := newPeerClient(s.opts.HubbleTarget, s.opts.DialTimeout)
+		if err != nil {
+			s.log.WithFields(logrus.Fields{
+				"error":        err,
+				"dial timeout": s.opts.DialTimeout,
+			}).Warning("Failed to create peer client for peers synchronization; will try again after the timeout has expired")
+			select {
+			case <-s.stop:
+				cancel()
+				return
+			case <-time.After(s.opts.RetryTimeout):
+				continue
+			}
+		}
+		client, err := cl.Notify(ctx, &peerpb.NotifyRequest{})
+		if err != nil {
+			conn.Close()
+			s.log.WithFields(logrus.Fields{
+				"error":              err,
+				"connection timeout": s.opts.RetryTimeout,
+			}).Warning("Failed to create peer notify client for peers change notification; will try again after the timeout has expired")
+			select {
+			case <-s.stop:
+				cancel()
+				return
+			case <-time.After(s.opts.RetryTimeout):
+				continue
+			}
+		}
+		for {
+			select {
+			case <-s.stop:
+				conn.Close()
+				cancel()
+				return
+			default:
+			}
+			cn, err := client.Recv()
+			if err != nil {
+				conn.Close()
+				s.log.WithFields(logrus.Fields{
+					"error":              err,
+					"connection timeout": s.opts.RetryTimeout,
+				}).Warning("Error while receiving peer change notification; will try again after the timeout has expired")
+				select {
+				case <-s.stop:
+					cancel()
+					return
+				case <-time.After(s.opts.RetryTimeout):
+					continue connect
+				}
+			}
+			s.log.WithField("change notification", cn).Debug("Received peer change notification")
+			p := peer.FromChangeNotification(cn)
+			switch cn.GetType() {
+			case peerpb.ChangeNotificationType_PEER_ADDED:
+				s.addPeer(p)
+			case peerpb.ChangeNotificationType_PEER_DELETED:
+				s.deletePeer(p)
+			case peerpb.ChangeNotificationType_PEER_UPDATED:
+				s.updatePeer(p)
+			}
+		}
+	}
+}
+
+func (s *Server) addPeer(p *peer.Peer) {
+	if p == nil {
+		return
+	}
+	s.mu.Lock()
+	s.peers[p.Name] = p
+	s.mu.Unlock()
+}
+
+func (s *Server) deletePeer(p *peer.Peer) {
+	if p == nil {
+		return
+	}
+	s.mu.Lock()
+	delete(s.peers, p.Name)
+	s.mu.Unlock()
+}
+
+func (s *Server) updatePeer(p *peer.Peer) {
+	if p == nil {
+		return
+	}
+	s.mu.Lock()
+	s.peers[p.Name] = p
+	s.mu.Unlock()
+}
+
+func (s *Server) peerList() []peer.Peer {
+	s.mu.Lock()
+	p := make([]peer.Peer, 0, len(s.peers))
+	for _, v := range s.peers {
+		if v != nil {
```

Comment:
out of curiosity: Since we already do nil checks in `{add,delete,update}Peer`, could we do the pointer de-reference there? Then this for-loop could b replaced by a single `copy` statement.

Reply:
This is a nice suggestion. However, the for-loop is still needed because it's converting a map to a slice so a `copy` operation won't work here.